### PR TITLE
Addresses issue #376

### PR DESCRIPTION
**Pull request by honorabrutroll**
Addresses issue #376 by checking the compilation mode before printing.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! I think you may actually be able to avoid passing around `CompileMode`, the recent refactoring means you've got access to a [`Key`](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/job_queue.rs#L41-L47) when calling [`note_working_on`](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/job_queue.rs#L184) which in turns has access to a [`Profile`](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/manifest.rs#L109-L121) which has a [field to indicate](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/manifest.rs#L119) whether it's documentation or not.

I think the logic here may want to instead be:

* If `doc` is true, print `Documenting` if it's not already been printed
* If `doc` is false, print `Compiling` if it's not already been printed

This may mean that some crates are printed as both `Documenting` and `Compiling`, but that's what's happening so seems fine to me!

**Comment by honorabrutroll** [comment]
Ah. Most of my changes were implemented before the refactoring and I just made them work, so I can probably get it to work with Profile instead now.

**Comment by honorabrutroll** [comment]
Using `Profile` should also allow the crates to print `Compiling` and just have the local crate print `Documenting`.

**Comment by honorabrutroll** [comment]
I got it to work with `Profile` instead and changed all of the relevant tests, and `test_cargo_doc::simple` ends up printing `Documenting` instead of `Compiling`, when it probably shouldn't.

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 1]
```diff
@@ -181,7 +181,7 @@ impl<'a> JobQueue<'a> {
         });
 
         // Print out some nice progress information
-        try!(self.note_working_on(config, key.pkg, fresh));
+        try!(self.note_working_on(config, key.pkg, fresh, &key.profile.doc));
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: This should actually be fine to just pass the whole `Key` into the function below and then it can decide which part to access

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 2]
```diff
@@ -210,7 +210,11 @@ impl<'a> JobQueue<'a> {
             // being a compiled package
             Dirty => {
                 self.printed.insert(pkg);
-                try!(config.shell().status("Compiling", pkg));
+                //check if documenting or compiling
+                match doc {
+                    &true => try!(config.shell().status("Documenting", pkg)),
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: I think that the logic may want to be beefed up a bit here. This means that each package *either* prints whether it's being documented or compiled, but it's also possible that both are done for a package (e.g. deps are both compiled and documented by default).

Perhaps the `printed` set could be split into `documented` and `compiled`, and then a package can both be printed as being documented and being compiled.

Comment (honorabrutroll) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: Wouldn't it be better to just keep `printed` and just do both printings at once?

Comment (honorabrutroll) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: Never mind about that first comment. So if I understand correctly, each dependency package would have both a compile job and a documentation job in the `JobQueue`?

**Comment by honorabrutroll** [comment]
The tests still pose a problem, as because the compile and documentation jobs run in parallel, each dependency crate's output can be of two results.

**Comment by honorabrutroll** [comment]
I've tried to implement a double or checker in my `dev` branch [here](https://github.com/honorabrutroll/cargo/blob/dev/tests/support/mod.rs#L554), but so far it hasn't worked correctly. Any input as to what might be wrong?

**Comment by alexcrichton** [comment]
> I got it to work with Profile instead and changed all of the relevant tests, and test_cargo_doc::simple ends up printing Documenting instead of Compiling, when it probably shouldn't.

Ah this is actually fine because some crates are actually both compiled and documented!

> The tests still pose a problem, as because the compile and documentation jobs run in parallel, each dependency crate's output can be of two results.

Hm that can indeed be a problem...

> I've tried to implement a double or checker in my dev branch here, but so far it hasn't worked correctly. Any input as to what might be wrong?

I'll take a look!

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 3]
```diff
@@ -199,9 +201,12 @@ impl<'a> JobQueue<'a> {
     // In general, we try to print "Compiling" for the first nontrivial task
     // run for a package, regardless of when that is. We then don't print
     // out any more information for a package after we've printed it once.
-    fn note_working_on(&mut self, config: &Config, pkg: &'a PackageId,
+    fn note_working_on(&mut self, config: &Config, key: &Key<'a>,
                        fresh: Freshness) -> CargoResult<()> {
-        if self.printed.contains(&pkg) {
+        let pkg: &'a PackageId = key.pkg;
+        let doc: &bool = &key.profile.doc;
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: It should be ok to not bind these to local variables, you can probably just use them through `key` directly.

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 4]
```diff
@@ -209,11 +214,16 @@ impl<'a> JobQueue<'a> {
...
             Dirty => {
-                self.printed.insert(pkg);
-                try!(config.shell().status("Compiling", pkg));
+                //check if documenting or compiling
+                match &key.profile.doc {
+                    &true => {self.documented.insert(pkg);
+                        try!(config.shell().status("Documenting", pkg));},
+                    _     => {self.compiled.insert(pkg);
+                        try!(config.shell().status("Compiling", pkg));},
+                }
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: Stylistically this'll probably be better formatted as:

```rust
if key.profile.doc {
    self.documented.insert(key.pkg);
    try!(config.shell().status("Documenting", key.pkg));
} else {
    // ...
}
```

**Comment by alexcrichton** [comment]
Ah you can probably avoid the custom function entirely and just not worry about the order that "Documented" and "Compiled" are printed in. You can just use `[..]` to ignore the two.

**Comment by honorabrutroll** [comment]
Oh ok, I didn't know that you could just ignore them like that.

**Comment by honorabrutroll** [comment]
@alexcrichton Could you kill and restart these CI builds? Looks like some of them have weird error and are holding up the queue.

**Comment by honorabrutroll** [comment]
The travis build stalling is probably related to [this](https://www.traviscistatus.com/incidents/frn0nyd2jhq7).

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 5]
```diff
@@ -209,12 +212,19 @@ impl<'a> JobQueue<'a> {
...
             // being a compiled package
             Dirty => {
-                self.printed.insert(pkg);
-                try!(config.shell().status("Compiling", pkg));
+                //check if documenting or compiling
+                if key.profile.doc {
+                    self.documented.insert(key.pkg);
+                    try!(config.shell().status("Documenting", key.pkg));
+                }
+                else {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: Stylistically we typically have the closing brace of the `if` and `else` on the same line

**Comment by alexcrichton** [comment]
Thanks @honorabrutroll! This looks good to me, could you also squash the commits down? (should trigger a new CI build as well)

**Comment by honorabrutroll** [comment]
How far back do you want me to squash them to? The first commit I assume?

**Comment by honorabrutroll** [comment]
@alexcrichton The travis builds are still stuck at build #3494 (the broken one).

File: Makefile.in [hunk 6]
```diff
@@ -34,10 +34,10 @@ else
 MAYBE_DISABLE_VERIFY=
 endif
 
-ifdef CFG_ENABLE_OPTIMIZE
-OPT_FLAG=--release
-else
+ifdef CFG_DISABLE_OPTIMIZE
```
Comment (alexcrichton) [inline Makefile.in]: Hm these seem like they may be unrelated changes, perhaps a rebase is necessary?

**Comment by honorabrutroll** [comment]
Hmm yeah, I'm new to squashing the commits so I just did 
`git reset --soft HEAD~12`
`git commit`
when i was squashing. I'll try rebasing it.

**Comment by honorabrutroll** [comment]
I think this looks about right?

**Comment by alexcrichton** [comment]
Ah yeah so what you'll want to do now is probably:

```
git fetch --all
git rebase origin/master
```

When you do a squash you probably want to do a `git rebase -i` instead of a reset + commit as it'll also allow you to reorder commits and such, but it'll also alert you if you go too far back as you'll see too much history. In general you only want to squash the commits you authored, not the previous merge commits.

**Comment by honorabrutroll** [comment]
Ah ok, I'll keep that in mind next time.

**Comment by alexcrichton** [comment]
(it looks like this may still need another rebase, there's a duplicate commit and a merge commit still here)

**Comment by honorabrutroll** [comment]
Done.

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 7]
```diff
@@ -209,12 +212,18 @@ impl<'a> JobQueue<'a> {
             // Any dirty stage which runs at least one command gets printed as
             // being a compiled package
             Dirty => {
-                self.printed.insert(pkg);
-                try!(config.shell().status("Compiling", pkg));
+                //check if documenting or compiling
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: Ah it's ok to leave out a comment like this, it pretty clearly follows from the code below!

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 8]
```diff
@@ -199,9 +201,10 @@ impl<'a> JobQueue<'a> {
     // In general, we try to print "Compiling" for the first nontrivial task
     // run for a package, regardless of when that is. We then don't print
     // out any more information for a package after we've printed it once.
-    fn note_working_on(&mut self, config: &Config, pkg: &'a PackageId,
+    fn note_working_on(&mut self, config: &Config, key: &Key<'a>,
                        fresh: Freshness) -> CargoResult<()> {
-        if self.printed.contains(&pkg) {
+        if (self.compiled.contains(key.pkg) && key.profile.doc == false) || 
+            (self.documented.contains(key.pkg) && key.profile.doc == true) {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: It's generally more idiomatic to just have `if foo` rather than `if foo == true` or `if !foo` instead of `if foo == false` in these two boolean checks.

**Comment by honorabrutroll** [comment]
By the way, you may want to delete some of the queued CI tests for this PR (there are ~5 of them queued).

**Comment by alexcrichton** [comment]
@bors: r+ 5d9ce21267a4585210a4d21be2367cce8c8d2c9e

Ah no worries, it'll catch up!

**Comment by alexcrichton** [comment]
Looks like there's some nondeterminism that was introduced which may need to be dealt with in the tests?

**Comment by honorabrutroll** [comment]
Strange that it only does that with homu and not appveyor...

**Comment by honorabrutroll** [comment]
I honestly don't how I would do this other than substituting `[..]` for all 3 outputs.

**Comment by alexcrichton** [comment]
It's probably fine to either tweak the test to not exhaustively match the output or to use the `with_stdout_contains` method to just match a subset of the output.

**Comment by alexcrichton** [comment]
@bors: r+ 925fb508504c3b8c23fe1c951cbd24e8c75ba96f

Let's see if it works this time!
